### PR TITLE
feat: add allocate_assignments method to content_assignments app.

### PR DESCRIPTION
**Pull request by iloveagent57**
ENT-7531

### CODE REVIEW

File: enterprise_access/apps/content_assignments/api.py [hunk 1]
```diff
@@ -21,28 +21,161 @@ def get_assignment_configuration(uuid):
...
+def allocate_assignments(assignment_configuration, learner_emails, content_key, content_price_cents):
+    """
+    Creates or updates an allocated assignment record
+    for the given ``content_key`` in the given ``assignment_configuration``,
+    for each email in the list of ``learner_emails``.
+
+    For existing assignment records with a (config, learner, content) combination, this function
+    does the following:
+      * If the existing record is cancelled or errored, update the existing record state to allocated
+      * If the existing record is allocated or accepted, don't do anything with the record
```
Comment (pwnage101) [inline enterprise_access/apps/content_assignments/api.py:81]: Non-blocking product question to ponder: if an existing record is allocated, should a notification be re-sent?

File: enterprise_access/apps/content_assignments/api.py [hunk 2]
```diff
@@ -21,28 +21,161 @@ def get_assignment_configuration(uuid):
...
+    # Get a list of refreshed objects that we just updated
+    return LearnerContentAssignment.objects.filter(
+        uuid__in=[record.uuid for record in assignment_records],
+    )
+
+
+def _create_new_assignments(assignment_configuration, learner_emails, content_key, content_price_cents):
+    """
+    Helper to bulk save new LearnerContentAssignment instances.
+    """
```
Comment (pwnage101) [inline enterprise_access/apps/content_assignments/api.py:168]: In the future, is this the function that also kicks off linking/notification tasks?

**Review by pwnage101 — APPROVED** [review]
the rest LGTM!

File: enterprise_access/apps/content_assignments/api.py (hunk 2 above)
Comment (iloveagent57) [reply enterprise_access/apps/content_assignments/api.py:168]: Yeah, either here, or do it directly from `allocated_assignments()` above, calling something like `notify_learner()` on the `updated_assignments` and `created_assignments` records.

File: enterprise_access/apps/content_assignments/models.py [hunk 3]
```diff
@@ -151,3 +158,44 @@ class Meta:
...
     )
     history = HistoricalRecords()
+
+    def __str__(self):
+        return (
+            f'uuid={self.uuid}, state={self.state}, learner_email={self.learner_email}, content_key={self.content_key}'
+        )
+
+    @classmethod
+    def bulk_create(cls, assignment_records):
```
[the remark this answers is not in the document]
Comment (iloveagent57) [reply enterprise_access/apps/content_assignments/models.py:177]: Ah good point, I'll change that name in an upcoming PR.

**Pull request merged**